### PR TITLE
MS Outlook new actions & sources #1576

### PR DESCRIPTION
**Pull request by feyzullah**
Added new actions;
* Update Contact
* Create Calendar Event
* Create Contact
* Create Draft Email
* Find Contact
* List Contacts
Refactored action:
* Send Email
Added new sources;
*New Email
*New Contact
*New Calendar Event
*Updated Calendar Event
Could not find proper way in API, could not add this source. If it is found I can add this too:
*Calendar Event Start

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

File: components/microsoft_outlook/actions/list-contacts/list-contacts.mjs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+      type: "string",
+      optional: true,
+    },
+  },
+  async run({ $ }) {
+    const response = await this.microsoftOutlook.listContacts({
+      $,
+      filterAddress: this.filterAddress,
+    });
+    $.export("$summary", `${response.value.length} contacts has been retrieved.`);
```
Comment (vellames-turing) [inline components/microsoft_outlook/actions/list-contacts/list-contacts.mjs:23]:
```suggestion
    $.export("$summary", `${response.value.length} contact(s) has been retrieved.`);
```
or you can use an if to check the phrase will be in the plural :)

**Review by vellames-turing — DISMISSED** [review]
Hi there, I left some minors. But looks great
Also, I saw some variables being repeated through the `actions`. Maybe we can move it to the `app`?
And I would like to suggest you to use `this.$emit` instead of returning values on `sources`

**Comment by DilanAthukorala** [comment]
@feyzullah **Observing the following issue for create-draft-email and send-email actions**

image



And for sources I didn't receive any events for new-contact and and new-email sources while testing.

File: components/microsoft_outlook/actions/find-contacts/find-contacts.mjs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    });
+    const relatedContacts = contactList.value.filter((c) => {
+      return c.displayName.includes(this.searchString) ||
+        c.givenName.includes(this.searchString) ||
+        c.surname.includes(this.searchString) ||
+        c.emailAddresses.find((e) =>
+          e.address == this.searchString ||
+          e.name.includes(this.searchString));
+    });
+    $.export("$summary", `${relatedContacts.length} contacts has been filtered.`);
```
[the remark this answers is not in the document]
Comment (feyzullah) [reply components/microsoft_outlook/actions/find-contacts/find-contacts.mjs]: @vellames-turing done, thanks!

File: components/microsoft_outlook/actions/list-contacts/list-contacts.mjs (hunk 1 above)
Comment (feyzullah) [reply components/microsoft_outlook/actions/list-contacts/list-contacts.mjs]: @vellames-turing done, thanks!

**Comment by feyzullah** [comment]
> 

Hi @DilanAthukorala thank you for finding it out, I guess I forgot to try with no file inputs. It must be resolved now, can you please check again?

**Comment by feyzullah** [comment]
> 

Hi @DilanAthukorala thank you for finding it out. I guess there was a bug related to nested `message` object. It was in send-email and not in create-draft-email, however I made a small change in create-draft-email too. It should have been fixed. Can you please check whenever you are available?

File: components/microsoft_outlook/microsoft_outlook.app.mjs [hunk 3]
```diff
@@ -14,49 +18,280 @@ export default {
...
+          value: tz.alias,
+        }));
+      },
+    },
+    contact: {
+      label: "Contact",
+      description: "The contact to be updated",
+      type: "string",
+      async options() {
+        const contactResponse = await this.listContacts();
```
Comment (jcortes) [inline components/microsoft_outlook/microsoft_outlook.app.mjs:56]: Shouldn't this one support pagination?

**Review by jcortes — CHANGES_REQUESTED** [review]
Hi @feyzullah I've just added a few minor observations before moving it to QA! Thanks

File: components/microsoft_outlook/microsoft_outlook.app.mjs (hunk 3 above)
Comment (feyzullah) [reply components/microsoft_outlook/microsoft_outlook.app.mjs:56]: Hi @jcortes tl;tr; it is not supported.
I double checked pagination [docs](https://docs.microsoft.com/en-us/graph/paging). It says that some of the resources does not support pagination. When I check [contacts](https://docs.microsoft.com/en-us/graph/api/user-list-contacts) it seems it is not supported and when I check [users](https://docs.microsoft.com/en-us/graph/api/user-list) as another example, it seems supported there. If you find more information please let me know. I left it as it was for now.

**Review by jcortes — APPROVED** [review]
LGTM! Ready for QA!

**Pull request merged**